### ISSUE TITLE
Segfault in neg, introduced in 3e6e81d

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#3412]
Received a segfault in backward of `neg` at commit 3c00c01 (at 8cd0df020ca6a251ab566e1c5a84627cbc70d483 there were no problems).

Confirmed, it is 3e6e81da460f6f90d65c0def58a3538cfe106cda that introduced this regression.

```
  read 39 "python" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffeff7fe700 (LWP 3823)]
0x00007fffc1523cd4 in at::Tensor::neg (this=0x7ffef00009f0)
    at ...torch/lib/tmp_install/include/ATen/TensorMethods.h:514
514         return type().neg(*this);
(gdb) bt
#0  0x00007fffc1523cd4 in at::Tensor::neg (this=0x7ffef00009f0)
    at .../torch/lib/tmp_install/include/ATen/TensorMethods.h:514
#1  torch::autograd::generated::NegBackward::apply (this=0x464afc00,
    grads=std::vector of length 1, capacity 1 = {...}) at torch/csrc/autograd/generated/Functions.cpp:1271
#2  0x00007fffc1472e6b in torch::autograd::Function::operator() (
    inputs=std::vector of length 1, capacity 1 = {...}, this=<optimized out>)
    at ...torch/csrc/autograd/function.h:89
#3  torch::autograd::call_function (task=...) at torch/csrc/autograd/engine.cpp:208
#4  torch::autograd::Engine::evaluate_function (this=this@entry=0x7fffc27764c0 <engine>, task=...)
    at torch/csrc/autograd/engine.cpp:220
#5  0x00007fffc1474fee in torch::autograd::Engine::thread_main (this=0x7fffc27764c0 <engine>, graph_task=0x0)
    at torch/csrc/autograd/engine.cpp:144
#6  0x00007fffc1471dd2 in torch::autograd::Engine::thread_init (this=this@entry=0x7fffc27764c0 <engine>,
    device=device@entry=-1) at torch/csrc/autograd/engine.cpp:121
#7  0x00007fffc149443a in torch::autograd::python::PythonEngine::thread_init (this=0x7fffc27764c0 <engine>,
    device=-1) at torch/csrc/autograd/python_engine.cpp:28
#8  0x00007fffae5ddc80 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#9  0x00007ffff7bc16ba in start_thread (arg=0x7ffeff7fe700) at pthread_create.c:333
#10 0x00007ffff78f73dd in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109

(gdb) print *this
$5 = {<at::detail::TensorBase> = {pImpl = 0x0}, <No data fields>}

(gdb) frame 1
#1  torch::autograd::generated::NegBackward::apply (this=0x464ab9d0, grads=std::vector of length 1, capacity 1 = {...})
    at torch/csrc/autograd/generated/Functions.cpp:1264
1264        grad_inputs[0] = grad.neg();
(gdb) print grad
$6 = <optimized out>
```

Is the fact that `grad` is optimized out causing the problem?

**Comment by colesbury** [comment]
@vadimkantorov, do you know what's using the output of `neg()`? I think some function is sending back a null gradient.

**Comment by vadimkantorov** [comment]
It’s quite strange. This happens when I use resnet18 as backbone. When I use AlexNet, it works fine.

Can I use the new tracing capability to find this out?

Otherwise, should there be null-checks? Better assertion error than segfault. Or is the perf penalty nasty?

Envoyé de mon téléphone Windows 10

De : Sam Gross
Envoyé le :mercredi 1 novembre 2017 16:18
À : pytorch/pytorch
Cc : Vadim Kantorov; Mention
Objet :Re: [pytorch/pytorch] Segfault in neg, introduced in 3e6e81d (#3412)

@vadimkantorov, do you know what's using the output of neg()? I think some function is sending back a null gradient.

**Comment by gchanan** [comment]
Could be another instance of: https://github.com/pytorch/pytorch/issues/3241

It would be great if you could help us track down the answer to @colesbury's question above; as you can see from https://github.com/pytorch/pytorch/issues/3241, the only case we've been able to reproduce this is on a fake example.

**Comment by vadimkantorov** [comment]
In the resnet18 version, I also use torch.distributions.Bernoulli to do a data-dependent drop-out. Probably that's the cause. I'll check it in a bit.

**Comment by vadimkantorov** [comment]
It is indeed the case, using the following function in the graph causes the segfault.
```
def dropout2(x, p = 1.0, training = False):
    if not training:
        return x
    prob = x.mul(p)
    noise = torch.distributions.Bernoulli(1 - prob).sample().div(1 - prob)
    return x * noise
```

**Comment by colesbury** [comment]
Fixed in #3433
